### PR TITLE
Extract keys crate

### PR DESCRIPTION
**Pull request by brson**
###  What have you changed?

Extract the `keys` module and `Key`/`Value`/`KvPair` definitons to a `keys` crate. Moving these out of tikv is a prerequisite to extracting `sst_importer`, which is a prerequisite to extracting `raftstore`, which is 30k loc.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

cargo test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

no

###  Does this PR affect `tidb-ansible`?

no

### CODE REVIEW

**Comment by siddontang** [comment]
Great work @brson 

PTAL @BusyJay

File: components/keys/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+edition = "2018"
+publish = false
+
+[dependencies]
+byteorder = "1.2"
+derive_more = "0.15.0"
+hex = "0.3"
+kvproto = { git = "https://github.com/pingcap/kvproto.git" }
+
+tikv_util = { path = "../tikv_util" }
```
Comment (siddontang) [inline components/keys/Cargo.toml]: do we need to depend tikv_alloc here?

**Comment by siddontang** [comment]
@zhouqiang-cl 

can we later support comparing compiling time?

File: components/keys/Cargo.toml (hunk 1 above)
Comment (brson) [reply components/keys/Cargo.toml]: there aren't any benchmarks in this crate so I don't think so

Comment (siddontang) [reply components/keys/Cargo.toml]: maybe it is better to add it?

File: components/keys/src/lib.rs [hunk 2]
```diff
@@ -229,6 +230,28 @@ pub fn data_end_key(region_end_key: &[u8]) -> Vec<u8> {
     }
 }
 
+#[derive(Debug, Display)]
+pub enum Error {
```
Comment (overvenus) [inline components/keys/src/lib.rs]: How about deriving `Fail`?

File: components/keys/src/lib.rs [hunk 3]
```diff
@@ -229,6 +230,28 @@ pub fn data_end_key(region_end_key: &[u8]) -> Vec<u8> {
...
+    #[display(fmt = "{} is not a valid raft log key", "hex::encode_upper(_0)")]
+    InvalidLogKey(Vec<u8>),
```
Comment (zhangjinpeng1987) [inline components/keys/src/lib.rs]: I prefer InvalidRaftLogKey.

File: components/keys/src/lib.rs (hunk 2 above)
Comment (breeswish) [reply components/keys/src/lib.rs]: Derive Fail + 1

**Comment by siddontang** [comment]
Friendly ping @brson

File: components/keys/Cargo.toml (hunk 1 above)
Comment (brson) [reply components/keys/Cargo.toml]: I have done so.

File: components/keys/src/lib.rs (hunk 2 above)
Comment (brson) [reply components/keys/src/lib.rs]: I have made this change on request. I don't see why it's desirable though.

File: components/keys/src/lib.rs (hunk 3 above)
Comment (brson) [reply components/keys/src/lib.rs]: Done.

**Comment by brson** [comment]
All comments addressed. PTAL @zhangjinpeng1987 @breeswish @overvenus

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987 @breeswish @overvenus

**Pull request merged**